### PR TITLE
Homework 03

### PR DESCRIPTION
**Pull request by YassineFnn**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -4,26 +4,100 @@ import Control.Monad (unless)
...
 
 instance Eq Expr where
-  (==) = undefined
+  NumExpr a == NumExpr b = a == b
+  SqrtExpr a == SqrtExpr b = a == b
+  AddExpr a b == AddExpr c d = (a==c) && (b==d)
+  SubExpr a b == SubExpr c d = (a==c) && (b==d)
+  MulExpr a b == MulExpr c d = (a==c) && (b==d)
+  DivExpr a b == DivExpr c d = (a==c) && (b==d)
+  PowerExpr a b == PowerExpr c d = (a==c) && (b==d)
```
Comment (kajigor) [inline Main.hs:31]: These parentheses are redundant.

File: Main.hs [hunk 2]
```diff
@@ -4,26 +4,100 @@ import Control.Monad (unless)
...
+  _ == _ = False
 
 data Error
+  = DivByZero
+  | NegativeSqrtNum
 
 instance Show Error where
-  show = undefined
+  show DivByZero = "Division by zero error"
+  show (NegativeSqrtNum) = "Cannot take square root of negative number"
```
Comment (kajigor) [inline Main.hs:40]: These parentheses are also redundant

File: Main.hs [hunk 3]
```diff
@@ -4,26 +4,100 @@ import Control.Monad (unless)
...
+            | b == 0 -> Left DivByZero
+            | otherwise -> Right (a / b)
+        (Left err, _) -> Left err
+        (_, Left err) -> Left err
+
+eval (PowerExpr a b) = 
+    case (eval a, eval b) of
+        (Right a, Right b) -> Right (a ** b)
+        (Left err, _) -> Left err
+        (_, Left err) -> Left err
```
Comment (kajigor) [inline Main.hs:86]: You have a lot of code repetition which can be abstracted away by introducing a helper function

File: Main.hs [hunk 4]
```diff
@@ -4,26 +4,100 @@ import Control.Monad (unless)
...
+  AddExpr a b == AddExpr c d = (a==c) && (b==d)
+  SubExpr a b == SubExpr c d = (a==c) && (b==d)
+  MulExpr a b == MulExpr c d = (a==c) && (b==d)
+  DivExpr a b == DivExpr c d = (a==c) && (b==d)
+  PowerExpr a b == PowerExpr c d = (a==c) && (b==d)
+  _ == _ = False
 
 data Error
+  = DivByZero
+  | NegativeSqrtNum
```
Comment (kajigor) [inline Main.hs:36]: It would be helpful if an error contained info about the sub-expression, which produced it.

**Review by kajigor — CHANGES_REQUESTED** [review]

File: Main.hs (hunk 3 above)
Comment (YassineFnn) [reply Main.hs]: Fixed

File: Main.hs (hunk 2 above)
Comment (YassineFnn) [reply Main.hs]: Fixed

File: Main.hs (hunk 1 above)
Comment (YassineFnn) [reply Main.hs]: Fixed

File: Main.hs (hunk 4 above)
Comment (YassineFnn) [reply Main.hs]: I think it's fixed now, I added some additional info for each sub-expression.

File: Main.hs [hunk 5]
```diff
@@ -4,26 +4,92 @@ import Control.Monad (unless)
...
+eval (SubExpr a b) = evalHelper (-) a b
+
+eval (MulExpr a b) = evalHelper (*) a b
+
+eval (PowerExpr a b) = evalHelper (**) a b
+
+eval (DivExpr a b) = 
+    case (eval a, eval b) of
+        (Right a, Right b) 
+            | b == 0 -> Left (DivByZero (NumExpr a) (NumExpr b))
```
[the remark this answers is not in the document]
Comment (YassineFnn) [reply Main.hs]: Fixed, and I also added a test case for it.

[the remark this answers is not in the document]
Comment (kajigor) [reply Main.hs]: `NumExpr (1 - 1)` is literally `0`. Have a look at your Expression representation. How do you represent subtraction?

[the remark this answers is not in the document]
Comment (YassineFnn) [reply Main.hs]: How about this way:
ghci> eval(DivExpr (NumExpr 10) (SubExpr (NumExpr 1) (NumExpr 1)))
Left Division by zero error: 10.0 / (1.0-1.0)

[the remark this answers is not in the document]
Comment (kajigor) [reply Main.hs]: :+1:

**Comment by kajigor** [comment]
Ok

**Pull request closed without merging**